tests/websockets/test_machine.py

```diff
@@ -36,6 +36,49 @@ async def test_unsuccessful_auth(self):
                 secure=False,
             )
 
+    async def test_successful_auth_api_key(self):
+        machine = await CachingMachine.create(
+            self._server.host,
+            api_key=self._server.api_key,
+            secure=False,
+        )
+        await machine.close()
+
+    async def test_unsuccessful_auth_api_key(self):
+        with self.assertRaises(websockets.exceptions.SecurityError):
+            await CachingMachine.create(
+                self._server.host,
+                api_key="not a real api_key",
+                secure=False,
+            )
+
+    async def test_unsuccessful_auth_api_key_and_password(self):
+        with self.assertRaises(ValueError):
+            await CachingMachine.create(
+                self._server.host,
+                username=self._server.username,
+                password=self._server.password,
+                api_key=self._server.api_key,
+                secure=False,
+            )
+
+    async def test_unsuccessful_auth_no_input(self):
+        with self.assertRaises(ValueError):
+            await CachingMachine.create(
+                self._server.host,
+                secure=False,
+            )
+
+    async def test_unsuccessful_auth_no_username(self):
+        # username has default value "root"
```

Comment:
If you make that default to `None` instead, this gets simpler.  I think we should make that change now that there are two different ways to authenticate that are supported.